src/jubjub/mod.rs

```diff
@@ -434,6 +446,257 @@ impl JubjubBls12 {
+    let mut offset = 4u64;
+
+    let window_size = {
+        file.seek(SeekFrom::Start(offset)).unwrap();
+        let mut buf = [0u8; 4];
+        file.read_exact(&mut buf).expect("failed to read bytes 4-8 from the exp-table file");
+        offset += 4;
+        u32::from_le_bytes(buf)
+    };
+
+    // Calculate the number of bytes we have to seek forward per segment in the exp-table file.
+    let n_windows_per_segment = {
+        let n_windows_per_segment = fs::Fs::NUM_BITS as f32 / window_size as f32;
+        if n_windows_per_segment.fract() == 0.0 {
+            (n_windows_per_segment as usize) + 1
+        } else {
+            n_windows_per_segment.ceil() as usize
+        }
+    };
+    let n_points_per_window = 2usize.pow(window_size);
+    let n_bytes_per_point = size_of::<Fr>();
+    let n_bytes_per_segment = n_windows_per_segment * n_points_per_window * n_bytes_per_point;
+
+    // Move the offset to the first byte of the first segment that we want to read.
+    offset += n_bytes_per_segment as u64 * first_segment as u64;
+    file.seek(SeekFrom::Start(offset)).unwrap();
+
+    // Read in the desired range of the exp-table.
+    let mut exp_table = vec![];
+
+    for _ in 0..n_segments {
+        let mut segment_windows = vec![];
+        for _ in 0..n_windows_per_segment {
+            let mut window = vec![];
+            for _ in 0..n_points_per_window {
+                // TODO: reduce the number of file reads.
+                let point = match edwards::Point::<Bls12, Unknown>::read(&file, &tmp_params) {
+                    Ok(point) => {
+                        point
+                            .as_prime_order(&tmp_params)
+                            .expect("point read from exp-table does not belong to subgroup")
+                    }
+                    Err(err) => {
+                        match err.kind() {
+                            io::ErrorKind::UnexpectedEof => {
+                                // If we have reached the end of the file and have successfully read
+                                // the last segment's table, return the exp-table. Otherwise, if we
+                                // have reached the end of the exp-table file in the middle of a
+                                // segment's table, panic.
+                                if segment_windows.is_empty() && window.is_empty() {
+                                    return exp_table;
+                                }
+                                panic!("exp-table file ends in the middle of a segment's table");
+                            }
+                            _ => {
+                                panic!("point read from exp-table does not exist on curve");
+                            }
+                        }
+                    }
+                };
+
+                window.push(point);
+                offset += n_bytes_per_point as u64;
+
+                file.seek(SeekFrom::Start(offset)).unwrap();
```

Comment:
why do you have to seek everytime? didn't you just read `n_bytes_per_point` starting at the old `offset`?